### PR TITLE
Provide ability for Worker Kubeadm Config Patches

### PR DESCRIPTION
**Pull request by robertojrojas**
Issue Reference #1084 

Hopefully this is a good approach, but I'm open to any changes.
One thing is, I wonder if it makes more sense to place **`KubeadmConfigPatches`** and **`KubeadmConfigPatchesJSON6902`** into a separate struct and embed.

TODO:
 
- Integration Testing (working on this)

### CODE REVIEW

**Comment by neolit123** [comment]
@robertojrojas 
was this discussed in an issue?

in the earlier iterations of the kind config it supported per-node patches, but it then moved to having patches per cluster.

how would the per-node patches interact with the cluster patches?
apply cluster patches to all nodes, then apply per-node patches?

**Comment by robertojrojas** [comment]
@neolit123 My apologies I should have created an issue first. I just created it #1084 
We can continue discussing there.

File: pkg/internal/cluster/create/actions/config/config.go [hunk 1]
```diff
@@ -85,12 +84,28 @@ func (a *Action) Execute(ctx *actions.ActionContext) error {
 	if err != nil {
 		return err
 	}
+
+	kubeadmConfigPlusPatches := func(node nodes.Node, configData kubeadm.ConfigData) func() error {
+		return func() error {
+			// determine if we're dealing with a worker node
+			// so we can include it's corresponding patches
+			includeWorkerPatches := !configData.ControlPlane
```
Comment (BenTheElder) [inline pkg/internal/cluster/create/actions/config/config.go]: Why are worker nodes being special cased?
Why not just include any node level patches?

File: pkg/internal/cluster/create/actions/config/config.go [hunk 2]
```diff
@@ -153,12 +165,21 @@ func removeMetadata(kustomized string) string {
 	)
 }
 
-func allPatchesFromConfig(cfg *config.Cluster) (patches []string, jsonPatches []config.PatchJSON6902) {
-	return cfg.KubeadmConfigPatches, cfg.KubeadmConfigPatchesJSON6902
+func allPatchesFromConfig(cfg *config.Cluster, includeWorkerPatches bool) (patches []string, jsonPatches []config.PatchJSON6902) {
+	patches, jsonPatches = cfg.KubeadmConfigPatches, cfg.KubeadmConfigPatchesJSON6902
+
+	if includeWorkerPatches {
+		for _, node := range cfg.Nodes {
```
Comment (BenTheElder) [inline pkg/internal/cluster/create/actions/config/config.go]: Why would ever node's patches be applied to all nodes? Why not just the patch for the current node?

File: pkg/internal/cluster/create/actions/config/config.go [hunk 3]
```diff
@@ -121,22 +134,21 @@ func (a *Action) Execute(ctx *actions.ActionContext) error {
...
 		return "", err
 	}
-	// fix all the patches to have name metadata matching the generated config
-	patches, jsonPatches := allPatchesFromConfig(cfg)
+
 	// apply patches
-	// TODO(bentheelder): this does not respect per node patches at all
-	// either make patches cluster wide, or change this
-	patched, err := patch.Patch(config, patches, jsonPatches)
+	patched, err := patch.Patch(cf, patches, jsonPatches)
```
Comment (BenTheElder) [inline pkg/internal/cluster/create/actions/config/config.go]: This is going to have surprising behavior in the current implementation. All merge patches will be applied and then all 6902 patches.

As a user I would instead expect all cluster global patches and then all node level patches

File: hack/update/generated.sh [hunk 4]
```diff
@@ -45,7 +45,7 @@ cd "${FAKE_REPOPATH}"
 
 # run the generators
 bin/deepcopy-gen -i ./pkg/internal/apis/config/ -O zz_generated.deepcopy --go-header-file hack/tools/boilerplate.go.txt
-bin/deepcopy-gen -i ./pkg/apis/config/v1alpha3 -O zz_generated.deepcopy --go-header-file hack/tools/boilerplate.go.txt
+bin/deepcopy-gen -i ./pkg/apis/config/v1alpha4 -O zz_generated.deepcopy --go-header-file hack/tools/boilerplate.go.txt
```
Comment (BenTheElder) [inline hack/update/generated.sh]: v1alpha3 should not be removed

File: pkg/apis/config/v1alpha4/types.go [hunk 5]
```diff
@@ -85,6 +85,28 @@ type Node struct {
 	// ExtraPortMappings describes additional port mappings for the node container
 	// binded to a host Port
 	ExtraPortMappings []PortMapping `yaml:"extraPortMappings,omitempty"`
+
+	// KubeadmConfigPatches are applied to the generated kubeadm config as
```
Comment (BenTheElder) [inline pkg/apis/config/v1alpha4/types.go]: These should document their relationship to the cluster wide patches, IE that they are applied next and are more specific.

File: pkg/internal/cluster/create/actions/config/config.go (hunk 1 above)
Comment (robertojrojas) [reply pkg/internal/cluster/create/actions/config/config.go]: The idea was to only apply the worker patches to workers and not to the control planes nodes, but this should also merge with the cluster level patches.

File: pkg/internal/cluster/create/actions/config/config.go (hunk 2 above)
Comment (robertojrojas) [reply pkg/internal/cluster/create/actions/config/config.go]: Yes, you're right. The patches should be applied to the current node. Looks like we would need to send the current node down to this function or the like.

File: pkg/internal/cluster/create/actions/config/config.go (hunk 3 above)
Comment (robertojrojas) [reply pkg/internal/cluster/create/actions/config/config.go]: Yes, I agree as a user I would expect the same.  I didn't think about it throughly. So, something like this maybe? Of course, I would need to provide patches separately and better code. wdyt?

```Go
clusterPatched, err := patch.Patch(cf, clusterPatches, clusterJSONPatches)
...
allPatched, err := patch.Patch(clusterPatched nodePatches, nodeJSONPatches)
...

```

File: hack/update/generated.sh (hunk 4 above)
Comment (robertojrojas) [reply hack/update/generated.sh]: so both v1alpha3 and v1alpha4 be update and verified, right?

File: pkg/apis/config/v1alpha4/types.go (hunk 5 above)
Comment (robertojrojas) [reply pkg/apis/config/v1alpha4/types.go]: Agreed. At first I thought about struct embedding, but I honestly don't know if it'll make the code and docs more confusing.

File: hack/update/generated.sh (hunk 4 above)
Comment (BenTheElder) [reply hack/update/generated.sh]: right

File: pkg/apis/config/v1alpha4/types.go (hunk 5 above)
Comment (BenTheElder) [reply pkg/apis/config/v1alpha4/types.go]: we just need a comment here about the behavior here (being applied on top of the others)

File: pkg/internal/cluster/create/actions/config/config.go (hunk 3 above)
Comment (BenTheElder) [reply pkg/internal/cluster/create/actions/config/config.go]: I think this works, I would add that we can skip running the second patch logic in the case that there are no patches, but we should always run the first patch (it also has the behavior of normalizing the output).

```go
patchedConfig, err := patch.Patch(cf, clusterPatches, clusterJSONPatches)
// handle error
if len(nodePatches) > 0 || len(nodeJSONPatches) > 0 {
  patchedConfig, err = patch.Patch(patchedConfig, nodePatches, nodeJSONPatches)
  // handle error
}
```

File: pkg/apis/config/v1alpha4/types.go (hunk 5 above)
Comment (robertojrojas) [reply pkg/apis/config/v1alpha4/types.go]: Cool beans! I'll work on these changes later today. I'll be integration testing this quite a bit as I'm working on an issue for [kube-bench](https://github.com/aquasecurity/kube-bench/issues/439) 😄

Comment (BenTheElder) [reply pkg/apis/config/v1alpha4/types.go]: thanks :-)

File: pkg/internal/cluster/create/actions/config/config.go (hunk 3 above)
Comment (robertojrojas) [reply pkg/internal/cluster/create/actions/config/config.go]: @BenTheElder  I ran into an interesting situation - **config.Node** vs **nodes.Node**.  As you know, the later is what I have access to within the **getKubeadmConfig** function, so how do I match the right **config.Node.'KubeadmConfigPatches'** patch(es) to apply it to the current  **nodes.Node** without having anything to correlate with?  I thought about the Role, but that's is not specific enough. The array indexes from nodes.Node(i) == cfg.Nodes[i] might do it, but is the index reliable in this case?  thoughts?

**Comment by BenTheElder** [comment]
The index of the nodes.Node[] is not reliable.
However the name of the node currently maps to the role and index in the
config.
This is brittle coupling wise and needs to go away in a future version, but
it is stable for now.

We can match the entries based on naming them again back to the named nodes.

**Comment by robertojrojas** [comment]
@BenTheElder I must be missing something because I don't see a **'name'** on the node (nodes.Node type) 
It's an interface and I do see a **String()**, is that what you mean?

**Comment by BenTheElder** [comment]
yes, the nodes.Node interface allows obtaining the actual node name with
String()
https://github.com/kubernetes-sigs/kind/blob/55dd46c27c8af766c5712212b8a759c7e5e208cb/pkg/cluster/nodes/types.go#L28

**Comment by robertojrojas** [comment]
Is this the format of the String() output?
https://github.com/kubernetes-sigs/kind/blob/55dd46c27c8af766c5712212b8a759c7e5e208cb/pkg/internal/cluster/providers/provider/common/namer_test.go#L43

**Comment by BenTheElder** [comment]
yes, if we're going to recompute the node names we should use the namer.

**Comment by robertojrojas** [comment]
yep, I saw that.

**Comment by robertojrojas** [comment]
@BenTheElder Please let me know what you think. In the meantime, I'm going to be testing it.

File: pkg/internal/cluster/create/actions/config/config.go [hunk 6]
```diff
@@ -158,7 +181,7 @@ func allPatchesFromConfig(cfg *config.Cluster) (patches []string, jsonPatches []
 }
 
 // writeKubeadmConfig writes the kubeadm configuration in the specified node
-func writeKubeadmConfig(logger log.Logger, cfg *config.Cluster, data kubeadm.ConfigData, node nodes.Node) error {
+func writeKubeadmConfig(cfg *config.Cluster, kubeadmConfig string, data kubeadm.ConfigData, node nodes.Node) error {
```
Comment (BenTheElder) [inline pkg/internal/cluster/create/actions/config/config.go]: this is wrong, we should not be taking in data and the kubeadm config.
we fix up the data below, but we've already generated the config

Comment (robertojrojas) [reply pkg/internal/cluster/create/actions/config/config.go]: Good catch! It turns out that there were some **"data"** configs that I missed. Fixing...

**Comment by robertojrojas** [comment]
@BenTheElder So far my integration tests are working fine.
 I did noticed the **apiVersion** changed from 
_**`kind.sigs.k8s.io`**/v1alpha3_ 
to
 _**`kind.x-k8s.io`**/v1alpha4_

**Comment by BenTheElder** [comment]
Yes, there's two supported config versions for this release, we'll put this in v1alpha4, we also needed to move domains because of revised guidelines around that.

**Comment by BenTheElder** [comment]
Sorry for the delay, I had to do some test-infra oncall (rotation over now) and sort out some Kubernetes things pre-code freeze, I'll go over this tonight and try to get v0.6.0 rolled up for tomorrow including this 😁

**Comment by robertojrojas** [comment]
Awesome sauce!

**Comment by BenTheElder** [comment]
/lgtm
/approve
Thanks for this!
I have some bikeshed, but we can iterate later :+)

**Comment by BenTheElder** [comment]
Fixes https://github.com/kubernetes-sigs/kind/issues/1084

**Pull request merged**